src/markdown_exec/formatters/jupyter.py

```diff
@@ -0,0 +1,253 @@
+"""Formatter for executing code using Jupyter kernels."""
+
+from __future__ import annotations
+
+import atexit
+import functools
+import os
+import platform
+import queue
+import time
+from typing import TYPE_CHECKING, Any, Callable
+
+from jupyter_client import KernelManager
+from jupyter_client.kernelspec import KernelSpecManager
+
+from markdown_exec.formatters.base import ExecutionError, base_format
+from markdown_exec.rendering import code_block
+
+if TYPE_CHECKING:
+    from jupyter_client.blocking.client import BlockingKernelClient
+
+_SHUTDOWN_TIMEOUT = 10
+
+# Store kernel managers by session and kernel name
+_kernel_managers: dict[str, dict[tuple[str | None, str], KernelManager]] = {}
+_kernel_clients: dict[str, dict[tuple[str | None, str], BlockingKernelClient]] = {}
+
+
+def _get_available_kernels() -> dict[str, dict[str, Any]]:
+    """Get information about all available Jupyter kernels.
+
+    Returns:
+        A dictionary mapping kernel names to their information including:
+        - display_name: Human readable name
+        - language: The programming language
```

Comment:
lint: Python-Markdown requires new lines before lists:

```suggestion
        A dictionary mapping kernel names to their information including:

        - display_name: Human readable name
        - language: The programming language
```